lib/ecto/schema.ex

```diff
@@ -2061,6 +2062,30 @@ defmodule Ecto.Schema do
     Module.put_attribute(mod, :struct_fields, {name, assoc})
   end
 
+  defp validate_default!(_, nil), do: :ok
+  defp validate_default!(_, []), do: :ok
```

Comment:
Do we need these clauses? They should also be handled by Ecto.Type.dump?

Reply:
So, I'm in doubt on this one: 

Without the `nil` clause, some custom types will raise on dump. An example is: Ecto.Repo.AutogenerateTest.ParameterizedTypePrefixedID

Without the `[]` clause, I get failures with test scenarios that rely on `embeds_many SchemaDefinedInTestFilesOnly`, as the compiler can't seem to find the module for the Ecto.Embedded parameterized type.

Reply:
I believe something could be done in this direction, but I probably need help to find the best approach

Reply:
Then it this case, you either need to pass a default or you need to change `Ecto.Repo.AutogenerateTest.ParameterizedTypePrefixedID` to accept a `nil` on dump. I.e. we are actually catching a bug. 👍 

For embeds, you should probably move the place where the check happens. Check it on `__field__` directly so we don't run it for embeds and the like.

Reply:
Managed to remove both of them, had to change some test schemas.